cirq/contrib/quirk/url_to_circuit.py

```diff
@@ -13,22 +13,86 @@
 # limitations under the License.
 import json
 import urllib.parse
-from typing import Any, List, Dict, Optional
+from typing import Any, List, Dict, Optional, Sequence, cast, TYPE_CHECKING, \
+    Iterable
 
-import cirq
+from cirq import devices, circuits, ops
 from cirq.contrib.quirk.cells import (
     Cell,
     CellMaker,
     CellMakerArgs,
     generate_all_quirk_cell_makers,
+    ExplicitOperationsCell,
 )
 
-
-def quirk_url_to_circuit(quirk_url: str) -> 'cirq.Circuit':
+if TYPE_CHECKING:
+    import cirq
+
+
+def quirk_url_to_circuit(
+        quirk_url: str,
+        *,
+        qubits: Optional[Sequence['cirq.Qid']] = None,
+        extra_recognized: Iterable['cirq.contrib.quirk.cells.CellMaker'] = ()
+) -> 'cirq.Circuit':
+    """Parses a Cirq circuit out of a Quirk URL.
+
+    Args:
+        quirk_url: The URL of a bookmarked Quirk circuit. It is not required
+            that the domain be "algassert.com/quirk". The only important part of
+            the URL is the fragment (the part after the #).
+        qubits: Qubits to use in the circuit. The length of the list must be
+            at least the number of qubits in the Quirk circuit (including unused
+            qubits). The maximum number of qubits in a Quirk circuit is 16.
+            This argument defaults to `cirq.LineQubit.range(16)` when not
+            specified.
+        extra_recognized: A list of non-standard Quirk cell makers. This can be
+            used to parse URLs that come from a modified version of Quirk that
+            includes gates that Quirk doesn't define. Each entry must be a
+            `cirq.contrib.quirk.cells.CellMaker`, which is a `NamedTuple` with
+            an `identifier` (the string that identifies the gate type), a `size`
+            (the height of the operation; the number of qubits it covers), and
+            a `maker` function which takes a
+            `cirq.contrib.quirk.cells.CellMakerArgs` and returns a
+            `cirq.Operation` or a `cirq.contrib.quirk.cells.Cell`. The cell is
+            more flexible (it can modify other cells in the same column before
+            producing operations).
```

Comment:
Move the explanation of the structure of cell makers to where `CellMaker` is defined. When we modify the definition of `CellMaker`, we're unlikely to remember this text needs to be updated, too.

Reply:
Done.